File: src/address.js
/**
 * Salvium Address Handling
 *
 * Supports all 18 address types across:
 * - 3 networks: mainnet, testnet, stagenet
 * - 2 formats: legacy (CryptoNote), carrot (CARROT)
 * - 3 types: standard, integrated, subaddress
 */

import {
  NETWORK,
  ADDRESS_TYPE,
  ADDRESS_FORMAT,
  PREFIX_MAP,
  KEY_SIZE,
  PAYMENT_ID_SIZE,
  ADDRESS_DATA_SIZE,
  getPrefix
} from './constants.js';
import { decodeAddress, encodeAddress } from './base58.js';
import {
  cnSubaddress,
  carrotSubaddress,
  generatePaymentId as genPaymentId
} from './subaddress.js';

/**
 * Result of parsing an address
 * @typedef {Object} ParsedAddress
 * @property {boolean} valid - Whether the address is valid
 * @property {string|null} network - Network type (mainnet, testnet, stagenet)
 * @property {string|null} format - Address format (legacy, carrot)
 * @property {string|null} type - Address type (standard, integrated, subaddress)
 * @property {string|null} prefix - Human-readable prefix (e.g., "SaLv", "SC1")
 * @property {Uint8Array|null} spendPublicKey - 32-byte public spend key
 * @property {Uint8Array|null} viewPublicKey - 32-byte public view key
 * @property {Uint8Array|null} paymentId - 8-byte payment ID (integrated addresses only)
 * @property {string|null} error - Error message if invalid
 */

/**
 * Parse and validate a Salvium address
 * @param {string} address - The address string to parse
 * @returns {ParsedAddress} - Parsed address information
 */
export function parseAddress(address) {
  const result = {
    valid: false,
    network: null,
    format: null,
    type: null,
    prefix: null,
    spendPublicKey: null,
    viewPublicKey: null,
    paymentId: null,
    error: null
  };

  // Basic validation
  if (!address || typeof address !== 'string') {
    result.error = 'Address must be a non-empty string';
    return result;
  }

  // Trim whitespace
  address = address.trim();

  // Check length (rough bounds based on possible address types)
  // Standard: ~95-99 chars, Integrated: ~106-110 chars, Subaddress: ~95-99 chars
  // Allow wide range to accommodate all prefix sizes
  if (address.length < 90 || address.length > 150) {
    result.error = 'Invalid address length';
    return result;
  }

  // Decode the address
  let decoded;
  try {
    decoded = decodeAddress(address);
  } catch (e) {
    result.error = e.message;
    return result;
  }

  const { tag, data } = decoded;

  // Look up the prefix
  const prefixInfo = PREFIX_MAP.get(tag);
  if (!prefixInfo) {
    result.error = `Unknown address prefix: 0x${tag.toString(16)}`;
    return result;
  }

  result.network = prefixInfo.network;
  result.format = prefixInfo.format;
  result.type = prefixInfo.type;
  result.prefix = prefixInfo.text;

  // Validate data length based on address type
  const expectedDataSize = ADDRESS_DATA_SIZE[result.type];
  if (data.length !== expectedDataSize) {
    result.error = `Invalid data length: expected ${expectedDataSize} bytes, got ${data.length}`;
    return result;
  }

  // Extract public keys
  result.spendPublicKey = data.slice(0, KEY_SIZE);
  result.viewPublicKey = data.slice(KEY_SIZE, KEY_SIZE * 2);

  // Extract payment ID for integrated addresses
  if (result.type === ADDRESS_TYPE.INTEGRATED) {
    result.paymentId = data.slice(KEY_SIZE * 2, KEY_SIZE * 2 + PAYMENT_ID_SIZE);
  }

  result.isCarrot = result.format === 'carrot';
  result.valid = true;
  return result;
}

/**
 * Validate a Salvium address
 * @param {string} address - The address string to validate
 * @returns {boolean} - True if valid, false otherwise
 */
export function isValidAddress(address) {
  return parseAddress(address).valid;
}

/**
 * Check if an address belongs to a specific network
 * @param {string} address - The address string
 * @param {string} network - Network to check (mainnet, testnet, stagenet)
 * @returns {boolean} - True if address belongs to the specified network
 */
export function isNetwork(address, network) {
  const parsed = parseAddress(address);
  return parsed.valid && parsed.network === network;
}

/**
 * Check if an address is a mainnet address
 * @param {string} address - The address string
 * @returns {boolean}
 */
export function isMainnet(address) {
  return isNetwork(address, NETWORK.MAINNET);
}

/**
 * Check if an address is a testnet address
 * @param {string} address - The address string
 * @returns {boolean}
 */
export function isTestnet(address) {
  return isNetwork(address, NETWORK.TESTNET);
}

/**
 * Check if an address is a stagenet address
 * @param {string} address - The address string
 * @returns {boolean}
 */
export function isStagenet(address) {
  return isNetwork(address, NETWORK.STAGENET);
}

/**
 * Check if an address uses the CARROT format
 * @param {string} address - The address string
 * @returns {boolean}
 */
export function isCarrot(address) {
  const parsed = parseAddress(address);
  return parsed.valid && parsed.format === ADDRESS_FORMAT.CARROT;
}

/**
 * Check if an address uses the legacy CryptoNote format
 * @param {string} address - The address string
 * @returns {boolean}
 */
export function isLegacy(address) {
  const parsed = parseAddress(address);
  return parsed.valid && parsed.format === ADDRESS_FORMAT.LEGACY;
}

/**
 * Check if an address is a standard address
 * @param {string} address - The address string
 * @returns {boolean}
 */
export function isStandard(address) {
  const parsed = parseAddress(address);
  return parsed.valid && parsed.type === ADDRESS_TYPE.STANDARD;
}

/**
 * Check if an address is an integrated address
 * @param {string} address - The address string
 * @returns {boolean}
 */
export function isIntegrated(address) {
  const parsed = parseAddress(address);
  return parsed.valid && parsed.type === ADDRESS_TYPE.INTEGRATED;
}

/**
 * Check if an address is a subaddress
 * @param {string} address - The address string
 * @returns {boolean}
 */
export function isSubaddress(address) {
  const parsed = parseAddress(address);
  return parsed.valid && parsed.type === ADDRESS_TYPE.SUBADDRESS;
}

/**
 * Get the public spend key from an address
 * @param {string} address - The address string
 * @returns {Uint8Array|null} - 32-byte public spend key or null if invalid
 */
export function getSpendPublicKey(address) {
  const parsed = parseAddress(address);
  return parsed.valid ? parsed.spendPublicKey : null;
}

/**
 * Get the public view key from an address
 * @param {string} address - The address string
 * @returns {Uint8Array|null} - 32-byte public view key or null if invalid
 */
export function getViewPublicKey(address) {
  const parsed = parseAddress(address);
  return parsed.valid ? parsed.viewPublicKey : null;
}

/**
 * Get the payment ID from an integrated address
 * @param {string} address - The address string
 * @returns {Uint8Array|null} - 8-byte payment ID or null if not an integrated address
 */
export function getPaymentId(address) {
  const parsed = parseAddress(address);
  return parsed.valid && parsed.type === ADDRESS_TYPE.INTEGRATED ? parsed.paymentId : null;
}

/**
 * Create an address from components
 * @param {Object} options - Address options
 * @param {string} options.network - Network type
 * @param {string} options.format - Address format
 * @param {string} options.type - Address type
 * @param {Uint8Array} options.spendPublicKey - 32-byte public spend key
 * @param {Uint8Array} options.viewPublicKey - 32-byte public view key
 * @param {Uint8Array} [options.paymentId] - 8-byte payment ID (for integrated addresses)
 * @returns {string|null} - Encoded address or null on error
 */
export function createAddress(options) {
  const { network, format, type, spendPublicKey, viewPublicKey, paymentId } = options;

  // Validate keys
  if (!spendPublicKey) {
    throw new Error('createAddress: spendPublicKey is required');
  }
  if (spendPublicKey.length !== KEY_SIZE) {
    throw new Error(`createAddress: spendPublicKey must be ${KEY_SIZE} bytes, got ${spendPublicKey.length}`);
  }
  if (!viewPublicKey) {
    throw new Error('createAddress: viewPublicKey is required');
  }
  if (viewPublicKey.length !== KEY_SIZE) {
    throw new Error(`createAddress: viewPublicKey must be ${KEY_SIZE} bytes, got ${viewPublicKey.length}`);
  }

  // Get prefix
  const prefix = getPrefix(network, format, type);
  if (prefix === null) {
    throw new Error(`createAddress: invalid network/format/type combination: ${network}/${format}/${type}`);
  }

  // Build data
  let data;
  if (type === ADDRESS_TYPE.INTEGRATED) {
    if (!paymentId) {
      throw new Error('createAddress: paymentId is required for integrated addresses');
    }
    if (paymentId.length !== PAYMENT_ID_SIZE) {
      throw new Error(`createAddress: paymentId must be ${PAYMENT_ID_SIZE} bytes, got ${paymentId.length}`);
    }
    data = new Uint8Array(KEY_SIZE * 2 + PAYMENT_ID_SIZE);
    data.set(spendPublicKey, 0);
    data.set(viewPublicKey, KEY_SIZE);
    data.set(paymentId, KEY_SIZE * 2);
  } else {
    data = new Uint8Array(KEY_SIZE * 2);
    data.set(spendPublicKey, 0);
    data.set(viewPublicKey, KEY_SIZE);
  }

  return encodeAddress(prefix, data);
}

/**
 * Convert a standard address to an integrated address by adding a payment ID
 * @param {string} address - Standard address
 * @param {Uint8Array|string} paymentId - 8-byte payment ID (or 16-char hex string)
 * @returns {string} - Integrated address
 * @throws {Error} If address is invalid or not a standard address
 */
export function toIntegratedAddress(address, paymentId) {
  const parsed = parseAddress(address);

  if (!parsed.valid) {
    throw new Error(`toIntegratedAddress: invalid address - ${parsed.error}`);
  }
  if (parsed.type !== ADDRESS_TYPE.STANDARD) {
    throw new Error(`toIntegratedAddress: address must be a standard address, got ${parsed.type}`);
  }

  // Convert hex string to bytes if needed
  if (typeof paymentId === 'string') {
    if (paymentId.length !== 16) {
      throw new Error(`toIntegratedAddress: payment ID hex string must be 16 characters, got ${paymentId.length}`);
    }
    const bytes = new Uint8Array(8);
    for (let i = 0; i < 8; i++) {
      bytes[i] = parseInt(paymentId.substr(i * 2, 2), 16);
    }
    paymentId = bytes;
  }

  if (paymentId.length !== PAYMENT_ID_SIZE) {
    throw new Error(`toIntegratedAddress: payment ID must be ${PAYMENT_ID_SIZE} bytes, got ${paymentId.length}`);
  }

  return createAddress({
    network: parsed.network,
    format: parsed.format,
    type: ADDRESS_TYPE.INTEGRATED,
    spendPublicKey: parsed.spendPublicKey,
    viewPublicKey: parsed.viewPublicKey,
    paymentId
  });
}

/**
 * Extract the standard address from an integrated address
 * @param {string} address - Integrated address
 * @returns {string} - Standard address
 * @throws {Error} If address is invalid or not an integrated address
 */
export function toStandardAddress(address) {
  const parsed = parseAddress(address);

  if (!parsed.valid) {
    throw new Error(`toStandardAddress: invalid address - ${parsed.error}`);
  }
  if (parsed.type !== ADDRESS_TYPE.INTEGRATED) {
    throw new Error(`toStandardAddress: address must be an integrated address, got ${parsed.type}`);
  }

  return createAddress({
    network: parsed.network,
    format: parsed.format,
    type: ADDRESS_TYPE.STANDARD,
    spendPublicKey: parsed.spendPublicKey,
    viewPublicKey: parsed.viewPublicKey
  });
}

/**
 * Format address info as a human-readable string
 * @param {string} address - The address string
 * @returns {string} - Human-readable description
 */
export function describeAddress(address) {
  const parsed = parseAddress(address);

  if (!parsed.valid) {
    return `Invalid address: ${parsed.error}`;
  }

  const parts = [
    parsed.network.charAt(0).toUpperCase() + parsed.network.slice(1),
    parsed.format === ADDRESS_FORMAT.CARROT ? 'CARROT' : 'Legacy',
    parsed.type
  ];

  if (parsed.type === ADDRESS_TYPE.INTEGRATED) {
    const paymentIdHex = Array.from(parsed.paymentId)
      .map(b => b.toString(16).padStart(2, '0'))
      .join('');
    parts.push(`(Payment ID: ${paymentIdHex})`);
  }

  return parts.join(' ');
}

/**
 * Convert bytes to hex string
 * @param {Uint8Array} bytes - Bytes to convert
 * @returns {string} - Hex string
 */
// Pre-computed hex lookup table — avoids Array.from().map().join() per call
const _hexLUT = /* @__PURE__ */ (() => {
  const t = new Array(256);
  for (let i = 0; i < 256; i++) t[i] = i.toString(16).padStart(2, '0');
  return t;
})();

export function bytesToHex(bytes) {
  let hex = '';
  for (let i = 0; i < bytes.length; i++) hex += _hexLUT[bytes[i]];
  return hex;
}

/**
 * Convert hex string to bytes
 * @param {string} hex - Hex string
 * @returns {Uint8Array} - Bytes
 */
export function hexToBytes(hex) {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(hex.substr(i * 2, 2), 16);
  }
  return bytes;
}

// ============================================================================
// Subaddress Generation
// ============================================================================

/**
 * Generate a CryptoNote (Legacy) subaddress
 * @param {Object} options - Subaddress options
 * @param {string} options.network - Network type (mainnet, testnet, stagenet)
 * @param {Uint8Array} options.spendPublicKey - 32-byte main spend public key
 * @param {Uint8Array} options.viewSecretKey - 32-byte view secret key
 * @param {number} options.major - Major index (account)
 * @param {number} options.minor - Minor index (address within account)
 * @returns {Object} { address, spendPublicKey, viewPublicKey }
 */
export function generateCNSubaddress(options) {
  const { network, spendPublicKey, viewSecretKey, major, minor } = options;

  // Generate subaddress keys
  const keys = cnSubaddress(spendPublicKey, viewSecretKey, major, minor);

  // Create the address string
  const address = createAddress({
    network,
    format: ADDRESS_FORMAT.LEGACY,
    type: ADDRESS_TYPE.SUBADDRESS,
    spendPublicKey: keys.spendPublicKey,
    viewPublicKey: keys.viewPublicKey
  });

  return {
    address,
    spendPublicKey: keys.spendPublicKey,
    viewPublicKey: keys.viewPublicKey,
    major,
    minor
  };
}

/**
 * Generate a CARROT subaddress
 * @param {Object} options - Subaddress options
 * @param {string} options.network - Network type (mainnet, testnet, stagenet)
 * @param {Uint8Array} options.accountSpendPubkey - K_s (32 bytes)
 * @param {Uint8Array} options.accountViewPubkey - K_v = k_vi × K_s (32 bytes)
 * @param {Uint8Array} options.generateAddressSecret - s_ga (32 bytes)
 * @param {number} options.major - Major index
 * @param {number} options.minor - Minor index
 * @returns {Object} { address, spendPublicKey, viewPublicKey, isMainAddress }
 */
export function generateCarrotSubaddress(options) {
  const { network, accountSpendPubkey, accountViewPubkey, generateAddressSecret, major, minor } = options;

  // Generate subaddress keys
  const keys = carrotSubaddress(accountSpendPubkey, accountViewPubkey, generateAddressSecret, major, minor);

  // For main address (0,0), use standard type; otherwise subaddress type
  const addressType = keys.isMainAddress ? ADDRESS_TYPE.STANDARD : ADDRESS_TYPE.SUBADDRESS;

  // Create the address string
  const address = createAddress({
    network,
    format: ADDRESS_FORMAT.CARROT,
    type: addressType,
    spendPublicKey: keys.spendPublicKey,
    viewPublicKey: keys.viewPublicKey
  });

  return {
    address,
    spendPublicKey: keys.spendPublicKey,
    viewPublicKey: keys.viewPublicKey,
    isMainAddress: keys.isMainAddress,
    major,
    minor
  };
}

/**
 * Generate a random 8-byte payment ID
 * @returns {Uint8Array} 8-byte payment ID
 */
export function generateRandomPaymentId() {
  return genPaymentId();
}

/**
 * Create an integrated address with a random payment ID
 * @param {string} address - Standard address
 * @returns {Object} { address, paymentId, paymentIdHex } - Integrated address and its payment ID
 * @throws {Error} If address is invalid or not a standard address
 */
export function createIntegratedAddressWithRandomId(address) {
  const paymentId = genPaymentId();
  const integratedAddress = toIntegratedAddress(address, paymentId);

  return {
    address: integratedAddress,
    paymentId,
    paymentIdHex: bytesToHex(paymentId)
  };
}

export default {
  parseAddress,
  isValidAddress,
  isNetwork,
  isMainnet,
  isTestnet,
  isStagenet,
  isCarrot,
  isLegacy,
  isStandard,
  isIntegrated,
  isSubaddress,
  getSpendPublicKey,
  getViewPublicKey,
  getPaymentId,
  createAddress,
  toIntegratedAddress,
  toStandardAddress,
  describeAddress,
  bytesToHex,
  hexToBytes,
  // Subaddress generation
  generateCNSubaddress,
  generateCarrotSubaddress,
  generateRandomPaymentId,
  createIntegratedAddressWithRandomId
};

File: crates/salvium-types/src/address.rs
//! Salvium address parsing, validation, and creation.
//!
//! Supports all 18 address types across 3 networks × 2 formats × 3 types.

use crate::base58;
use crate::constants::{
    address_data_size, prefix_info, get_prefix,
    AddressFormat, AddressType, Network,
    KEY_SIZE, PAYMENT_ID_SIZE,
};
use thiserror::Error;

#[derive(Debug, Error)]
pub enum AddressError {
    #[error("address must be a non-empty string")]
    Empty,

    #[error("invalid address length ({0})")]
    InvalidLength(usize),

    #[error("base58 decode error: {0}")]
    Base58(#[from] base58::Base58Error),

    #[error("unknown address prefix: 0x{0:x}")]
    UnknownPrefix(u64),

    #[error("invalid data length: expected {expected} bytes, got {actual}")]
    InvalidDataLength { expected: usize, actual: usize },

    #[error("spendPublicKey is required")]
    MissingSpendKey,

    #[error("viewPublicKey is required")]
    MissingViewKey,

    #[error("key must be {expected} bytes, got {actual}")]
    InvalidKeySize { expected: usize, actual: usize },

    #[error("paymentId is required for integrated addresses")]
    MissingPaymentId,

    #[error("paymentId must be {expected} bytes, got {actual}")]
    InvalidPaymentIdSize { expected: usize, actual: usize },

    #[error("invalid network/format/type combination")]
    InvalidCombination,

    #[error("address must be a standard address, got {0:?}")]
    NotStandard(AddressType),

    #[error("address must be an integrated address, got {0:?}")]
    NotIntegrated(AddressType),
}

/// Result of parsing an address.
#[derive(Debug, Clone)]
pub struct ParsedAddress {
    pub network: Network,
    pub format: AddressFormat,
    pub address_type: AddressType,
    pub prefix_text: &'static str,
    pub spend_public_key: [u8; KEY_SIZE],
    pub view_public_key: [u8; KEY_SIZE],
    pub payment_id: Option<[u8; PAYMENT_ID_SIZE]>,
}

impl ParsedAddress {
    pub fn is_carrot(&self) -> bool {
        self.format == AddressFormat::Carrot
    }

    pub fn is_legacy(&self) -> bool {
        self.format == AddressFormat::Legacy
    }

    /// Re-encode this parsed address back to a Base58 string.
    pub fn to_address_string(&self) -> String {
        create_address_raw(
            self.network,
            self.format,
            self.address_type,
            &self.spend_public_key,
            &self.view_public_key,
            self.payment_id.as_ref().map(|p| p.as_slice()),
        )
        .expect("ParsedAddress should always re-encode successfully")
    }
}

/// Parse and validate a Salvium address string.
pub fn parse_address(address: &str) -> Result<ParsedAddress, AddressError> {
    let address = address.trim();
    if address.is_empty() {
        return Err(AddressError::Empty);
    }

    // Rough length bounds (standard ~95-99, integrated ~106-110)
    if address.len() < 90 || address.len() > 150 {
        return Err(AddressError::InvalidLength(address.len()));
    }

    let (tag, data) = base58::decode_address(address)?;

    let info = prefix_info(tag).ok_or(AddressError::UnknownPrefix(tag))?;

    let expected_size = address_data_size(info.address_type);
    if data.len() != expected_size {
        return Err(AddressError::InvalidDataLength {
            expected: expected_size,
            actual: data.len(),
        });
    }

    let mut spend_public_key = [0u8; KEY_SIZE];
    spend_public_key.copy_from_slice(&data[..KEY_SIZE]);

    let mut view_public_key = [0u8; KEY_SIZE];
    view_public_key.copy_from_slice(&data[KEY_SIZE..KEY_SIZE * 2]);

    let payment_id = if info.address_type == AddressType::Integrated {
        let mut pid = [0u8; PAYMENT_ID_SIZE];
        pid.copy_from_slice(&data[KEY_SIZE * 2..KEY_SIZE * 2 + PAYMENT_ID_SIZE]);
        Some(pid)
    } else {
        None
    };

    Ok(ParsedAddress {
        network: info.network,
        format: info.format,
        address_type: info.address_type,
        prefix_text: info.text,
        spend_public_key,
        view_public_key,
        payment_id,
    })
}

/// Validate a Salvium address string.
pub fn is_valid_address(address: &str) -> bool {
    parse_address(address).is_ok()
}

/// Create an address string from components.
pub fn create_address_raw(
    network: Network,
    format: AddressFormat,
    addr_type: AddressType,
    spend_public_key: &[u8],
    view_public_key: &[u8],
    payment_id: Option<&[u8]>,
) -> Result<String, AddressError> {
    if spend_public_key.len() != KEY_SIZE {
        return Err(AddressError::InvalidKeySize {
            expected: KEY_SIZE,
            actual: spend_public_key.len(),
        });
    }
    if view_public_key.len() != KEY_SIZE {
        return Err(AddressError::InvalidKeySize {
            expected: KEY_SIZE,
            actual: view_public_key.len(),
        });
    }

    let prefix = get_prefix(network, format, addr_type)
        .ok_or(AddressError::InvalidCombination)?;

    let data = match addr_type {
        AddressType::Integrated => {
            let pid = payment_id.ok_or(AddressError::MissingPaymentId)?;
            if pid.len() != PAYMENT_ID_SIZE {
                return Err(AddressError::InvalidPaymentIdSize {
                    expected: PAYMENT_ID_SIZE,
                    actual: pid.len(),
                });
            }
            let mut d = Vec::with_capacity(KEY_SIZE * 2 + PAYMENT_ID_SIZE);
            d.extend_from_slice(spend_public_key);
            d.extend_from_slice(view_public_key);
            d.extend_from_slice(pid);
            d
        }
        _ => {
            let mut d = Vec::with_capacity(KEY_SIZE * 2);
            d.extend_from_slice(spend_public_key);
            d.extend_from_slice(view_public_key);
            d
        }
    };

    Ok(base58::encode_address(prefix, &data))
}

/// Convert a standard address to an integrated address by adding a payment ID.
pub fn to_integrated_address(
    address: &str,
    payment_id: &[u8; PAYMENT_ID_SIZE],
) -> Result<String, AddressError> {
    let parsed = parse_address(address)?;
    if parsed.address_type != AddressType::Standard {
        return Err(AddressError::NotStandard(parsed.address_type));
    }

    create_address_raw(
        parsed.network,
        parsed.format,
        AddressType::Integrated,
        &parsed.spend_public_key,
        &parsed.view_public_key,
        Some(payment_id),
    )
}

/// Extract the standard address from an integrated address.
pub fn to_standard_address(address: &str) -> Result<String, AddressError> {
    let parsed = parse_address(address)?;
    if parsed.address_type != AddressType::Integrated {
        return Err(AddressError::NotIntegrated(parsed.address_type));
    }

    create_address_raw(
        parsed.network,
        parsed.format,
        AddressType::Standard,
        &parsed.spend_public_key,
        &parsed.view_public_key,
        None,
    )
}

/// Describe an address in human-readable form.
pub fn describe_address(address: &str) -> String {
    match parse_address(address) {
        Ok(parsed) => {
            let network = match parsed.network {
                Network::Mainnet => "Mainnet",
                Network::Testnet => "Testnet",
                Network::Stagenet => "Stagenet",
            };
            let format = match parsed.format {
                AddressFormat::Legacy => "Legacy",
                AddressFormat::Carrot => "CARROT",
            };
            let addr_type = match parsed.address_type {
                AddressType::Standard => "standard",
                AddressType::Integrated => "integrated",
                AddressType::Subaddress => "subaddress",
            };

            if let Some(pid) = &parsed.payment_id {
                format!(
                    "{} {} {} (Payment ID: {})",
                    network, format, addr_type, hex::encode(pid)
                )
            } else {
                format!("{} {} {}", network, format, addr_type)
            }
        }
        Err(e) => format!("Invalid address: {}", e),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_mainnet_address_roundtrip() {
        // Create a mainnet legacy standard address with known keys
        let spend_key = [0x01u8; 32];
        let view_key = [0x02u8; 32];

        let address = create_address_raw(
            Network::Mainnet,
            AddressFormat::Legacy,
            AddressType::Standard,
            &spend_key,
            &view_key,
            None,
        )
        .unwrap();

        // Parse it back
        let parsed = parse_address(&address).unwrap();
        assert_eq!(parsed.network, Network::Mainnet);
        assert_eq!(parsed.format, AddressFormat::Legacy);
        assert_eq!(parsed.address_type, AddressType::Standard);
        assert_eq!(parsed.spend_public_key, spend_key);
        assert_eq!(parsed.view_public_key, view_key);
        assert!(parsed.payment_id.is_none());
    }

    #[test]
    fn test_integrated_address_roundtrip() {
        let spend_key = [0x11u8; 32];
        let view_key = [0x22u8; 32];
        let payment_id = [0xAA; PAYMENT_ID_SIZE];

        let address = create_address_raw(
            Network::Testnet,
            AddressFormat::Carrot,
            AddressType::Integrated,
            &spend_key,
            &view_key,
            Some(&payment_id),
        )
        .unwrap();

        let parsed = parse_address(&address).unwrap();
        assert_eq!(parsed.network, Network::Testnet);
        assert_eq!(parsed.format, AddressFormat::Carrot);
        assert_eq!(parsed.address_type, AddressType::Integrated);
        assert_eq!(parsed.payment_id, Some(payment_id));
    }

    #[test]
    fn test_integrated_conversion() {
        let spend_key = [0x33u8; 32];
        let view_key = [0x44u8; 32];

        let standard = create_address_raw(
            Network::Mainnet,
            AddressFormat::Legacy,
            AddressType::Standard,
            &spend_key,
            &view_key,
            None,
        )
        .unwrap();

        let payment_id = [0xBB; PAYMENT_ID_SIZE];
        let integrated = to_integrated_address(&standard, &payment_id).unwrap();
        let back = to_standard_address(&integrated).unwrap();

        let parsed_std = parse_address(&standard).unwrap();
        let parsed_back = parse_address(&back).unwrap();
        assert_eq!(parsed_std.spend_public_key, parsed_back.spend_public_key);
        assert_eq!(parsed_std.view_public_key, parsed_back.view_public_key);
    }
}
